### ISSUE TITLE
Customize segmentation model architecture to use detection model weights

### ISSUE THREAD

**Issue by rajas1310** [ultralytics/ultralytics#10237]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,
Thank you all for the amazing work! 
I have a trained a yolov8L model to detect humans in sport videos and it works very well. I aim to create a segmentation model using the body from my detection-model and the head from your pretrained segmentation model weights. I think the intermediate representations from my custom detection-model will enhance the segmentation output generated by the segmentation head.

I was able to load the detection weights from sub-modules 0 to 21 on the backbone architecture and your pretrained weights associated with the 22nd module from 'yolov8l-seg.pt'. But when I infer an image on this model, it gives masks = `None`, and empty lists for boxes attribute as well. 

The script I used to load the weights:
```
seg_model = YOLO('yolov8l-seg.pt')  # load an official model
det_model = YOLO('custom_human_detection.pt')

for i in range(22):
     seg_model.model.model[i].load_state_dict(det_model.model.model[i].state_dict())
```

The inference script:
```
results = seg_model.predict(img_path, max_det=1, verbose=False)  # predict on an image
print(results[0].masks)
p = results[0].plot(boxes=False)

try:
    if cv2.imwrite(f'image/saved/at/{im_idx}.jpg', p):
except Exception as e:
    print(e)
```
Output >>>
`None`

Could you please help me figure out what can be the problem?

### Additional

I recently checked the number of modules and parameters for my customized model (detection-body + segmentation-head) and I saw that before I run the inference the number of params in the pretrained segmentation model (L variant) is same to the number of params in my segmentation model. But after I run the Inference, the number of parameters in my customized model reduced by times and consequentially the number of params decrease as well.

Before inference
![Before inference](https://github.com/ultralytics/ultralytics/assets/56029517/8ee21d17-ffb8-445b-86fa-aa3dac7d5997)

After inference
![After inference](https://github.com/ultralytics/ultralytics/assets/56029517/5e348cd6-4b7e-41d7-a013-ac45f7b61f84)

**Comment by glenn-jocher** [comment]
Hello there 👋,

Thanks for reaching out and sharing your approach! It's great to hear about your success with human detection in sports videos using a YOLOv8 model.

From your description, it seems like you're trying to transfer learned features from a detection model to enhance a segmentation model's performance. This is indeed a feasible and exciting approach. However, the direct parameter copying you've performed might miss crucial steps necessary for proper initialization and functionality of the segmentation layers, particularly given the difference in output formats (bounding boxes vs. masks).

In the code snippet you provided for loading the weights, you've copied the state dicts successfully for modules 0 through 21. However, for segmentation tasks, ensuring the segmentation-specific layers are adequately initialized with the correct architectures and pretrained weights, as well as aligning input-output dimensions and feature spaces between the detection and segmentation parts, is critical.

As for the reduction of parameters post-inference and the issue with masks resulting in `None`, it's indicative of a possible misconfiguration or incompatibility between the model architectures or an issue in how the model state is being manipulated. Without a deep dive into the specifics of your customized architectures and the models' state dicts, it's challenging to pinpoint the exact issue.

One immediate suggestion is to carefully review the model layers responsible for outputting the masks and ensure they are being correctly utilized and that there's no resizing or transformation that alters their expected output shape. Additionally, consider whether any layers beyond the 22nd module require specific initialization or state_dict copying to maintain the expected inference behavior and output dimensions.

Unfortunately, directly merging models for different tasks like detection and segmentation requires careful adjustments beyond parameter sharing, particularly around layer compatibility and overall architecture design.

If you're eager to dive deeper into creating a hybrid model, exploring the model definition files and possibly extending the architectures with custom modules to bridge detection and segmentation tasks more seamlessly might be the way forward.

In the meantime, if you have more specifics or can share error logs or more detailed insights into the model architectures, I'd be happy to provide more targeted suggestions.

Keep up the awesome work! 🚀
